discovery/reliable_sender.go

```diff
@@ -0,0 +1,319 @@
+	// that should be reliably sent to peers while also taking into account
+	// the peer's connection lifecycle.
+	activePeers    map[[33]byte]peerManager
+	activePeersMtx sync.Mutex
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// newReliableSender returns a new reliableSender backed by the given config.
+func newReliableSender(cfg *reliableSenderCfg) *reliableSender {
+	return &reliableSender{
+		cfg:         *cfg,
+		activePeers: make(map[[33]byte]peerManager),
+		quit:        make(chan struct{}),
+	}
+}
+
+// Start spawns message handlers for any peers with pending messages.
+func (s *reliableSender) Start() error {
+	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
+		return nil
+	}
+
+	return s.resendPendingMsgs()
+}
+
+// Stop halts the reliable sender from sending messages to peers.
+func (s *reliableSender) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.stopped, 0, 1) {
+		return
+	}
+
+	close(s.quit)
+	s.wg.Wait()
+}
+
+// sendMessage constructs a request to send a message reliably to a peer. In the
+// event that the peer is currently offline, this will only write the message to
+// disk. Once the peer reconnects, this message, along with any others pending,
+// will be sent to the peer.
+func (s *reliableSender) sendMessage(msg lnwire.Message, peerPubKey [33]byte) error {
+	// We'll start by persisting the message to disk. This allows us to
+	// resend the message upon restarts and peer reconnections.
+	if err := s.cfg.MessageStore.AddMessage(msg, peerPubKey); err != nil {
+		return err
+	}
+
+	// Then, we'll spawn a peerHandler for this peer to handle resending its
+	// pending messages while taking into account its connection lifecycle.
+spawnHandler:
+	msgHandler, ok := s.spawnPeerHandler(peerPubKey)
+
+	// If the handler wasn't previously active, we can exit now as we know
+	// that the message will be sent once the peer online notification is
+	// received. This prevents us from potentially sending the message
+	// twice.
+	if !ok {
+		return nil
+	}
+
+	// Otherwise, we'll attempt to stream the message to the handler.
+	// There's a subtle race condition where the handler can be torn down
+	// due to all of the messages sent being stale, so we'll handle this
+	// gracefully by spawning another one to prevent blocking.
+	select {
+	case msgHandler.msgs <- msg:
+	case <-msgHandler.done:
+		goto spawnHandler
+	case <-s.quit:
+		return ErrGossiperShuttingDown
+	}
+
+	return nil
+}
+
+// spawnPeerMsgHandler spawns a peerHandler for the given peer if there isn't
+// one already active. The boolean returned signals whether there was already
+// one active or not.
+func (s *reliableSender) spawnPeerHandler(peerPubKey [33]byte) (peerManager, bool) {
+	s.activePeersMtx.Lock()
+	defer s.activePeersMtx.Unlock()
+
+	msgHandler, ok := s.activePeers[peerPubKey]
+	if !ok {
+		msgHandler = peerManager{
+			msgs: make(chan lnwire.Message),
+			done: make(chan struct{}),
+		}
+		s.activePeers[peerPubKey] = msgHandler
+
+		s.wg.Add(1)
```

Comment:
this can panic when calling `Wait` since this `Add` is done outside of a waitgrouped goroutine and can be called externally. in this case though, we can prevent this from happening by having the `reliableSender` shutdown after all of the gossiper's goroutines have exited, since that means no further calls `spawnPeerHandler` can happen, see my other comment

Reply:
Fixed.